### PR TITLE
Create Article menu item: displays error if Category is not selected when Default Category is set

### PR DESCRIPTION
**Pull request by infograf768**
Pull Request for Issue https://github.com/joomla/joomla-cms/issues/17708

### Summary of Changes
as title says

### Testing Instructions
Create a new menu item of type "Create Article"
In the Options tab, set Default Category to Yes
Do NOT select a category in the field below which only displays by a showon.
Save menu item.

### Before patch
The menu item is saved. No error is set. 
Category id was wrongly set to 1, which wrong behaviour is now corrected in frontend, among other issues with permissions, with https://github.com/joomla/joomla-cms/pull/17674 

### After patch
Preventing saving and displaying an error message

Screen Shot 2019-08-09 at 09 50 28

@alikon @LivioCavallo

### CODE REVIEW

**Comment by brianteeman** [comment]
1. I'm assuming that the obvious option of making the field required is not used for a valid reason

2. COM_CONTENT_CREATE_ARTICLE_ERROR="When default category is enabled, a category should be selected."
As its an error message not a help doc this can be rewritten to something like
COM_CONTENT_CREATE_ARTICLE_ERROR="A default category is required."

**Comment by infograf768** [comment]
>    I'm assuming that the obvious option of making the field required is not used for a valid reason

It's impossible as the field should be ignored when Default category is set to no and we do not have a specific "requiredon" (on the model of "showon" code).

>COM_CONTENT_CREATE_ARTICLE_ERROR="When default category is enabled, a category should be selected."
    As its an error message not a help doc this can be rewritten to something like
    COM_CONTENT_CREATE_ARTICLE_ERROR="A default category is required."

Why not give this detail? It does not cost much and help user solve the issue. grrr

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on 0a9638aa68ce41c3607f8801ee0fdc4575a43e5b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25798">issues.joomla.org/tracker/joomla-cms/25798</a>.</sub>

File: administrator/language/en-GB/en-GB.com_content.ini [hunk 1]
```diff
@@ -34,6 +34,7 @@ COM_CONTENT_CREATE_ARTICLE_CATEGORY_LABEL="Default Category"
 COM_CONTENT_CREATE_ARTICLE_CATEGORY_DESC="If set to 'Yes', this page will only let you create articles in the category selected below."
 COM_CONTENT_CREATE_ARTICLE_CUSTOM_CANCEL_REDIRECT_DESC="If set to 'Yes', you can set a redirection page, distinct from above 'Submission/Cancel Redirect', to redirect to when user Cancels article submission.<br />If set to 'No', when user Cancels article submission, the user is redirected to the above 'Submission/Cancel Redirect' page."
 COM_CONTENT_CREATE_ARTICLE_CUSTOM_CANCEL_REDIRECT_LABEL="Custom Redirect on Cancel"
+COM_CONTENT_CREATE_ARTICLE_ERROR="When default category is enabled, a category should be selected."
```
Comment (richard67) [inline administrator/language/en-GB/en-GB.com_content.ini:37]:
```suggestion
COM_CONTENT_CREATE_ARTICLE_ERROR="When default category is enabled, a category has to be selected."
```
"Should" is just a kind of recommendation, not a "must" or a "has to be" or a "needs to".

Comment (infograf768) [reply administrator/language/en-GB/en-GB.com_content.ini:37]: agree. merging now

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 261fdbe7e5012786417185022f96dede0a635111<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25798">issues.joomla.org/tracker/joomla-cms/25798</a>.</sub>

**Comment by richard67** [comment]
@franz-wohlkoenig Last change after your test was just a language change, so you can just set your test result again without retest, just review last change, and then RTC :smile:

**Comment by infograf768** [comment]
Don't worry about appveyor btw.

**Comment by richard67** [comment]
I know.

**Comment by ghost** [comment]
Status "Ready To Commit".

**Comment by HLeithner** [comment]
Sorry to say but this solution is really hacky in my opinion that should be done in a plugin for example content->joomla in the on_before_save event.

**Comment by infograf768** [comment]
@HLeithner 
I would not know how to do what you suggest.

**Comment by richard67** [comment]
@HLeithner I would also not know. I have an idea what you mean but I'm not good enough in event based stff to do it either. And it has 2 good tests, one from me. And I would not call it "hacky", only maybe "old style but working". So for me it's pretty fine.

P.S.: And it fixes an issue.

**Comment by HLeithner** [comment]
@richard67 writing a hardcoded string into com_menu because a specific other component needs it is wrong. It's possible that we don't have another way but if we have the choice we should use it.

Please but this in to plg_content_joomla + the language string.
```
	/**
	 * The save event.
	 *
	 * @param   string   $context  The context
	 * @param   object   $table    The item
	 * @param   boolean  $isNew    Is new item
	 * @param   array    $data     The validated data
	 *
	 * @return  void
	 *
	 * @since   __DEPLOY_VERSION__
	 */
	public function onContentBeforeSave($context, $table, $isNew, $data)
	{
		// Check we are handling the frontend edit form.
		if ($context !== 'com_menus.item')
		{
			return true;
		}

		// Special case for Create article menu item
		if ($table->link !== 'index.php?option=com_content&view=form&layout=edit')
		{
			return true;
		}

		// Display error if catid is not set when enable_category is enabled
		$params = json_decode($table->params, true);

		if ($params['enable_category'] == 1 && empty($params['catid']))
		{
			$table->setError(JText::_('PLG_CONTENT_JOOMLA_CREATE_ARTICLE_WITH_EMPTY_CATEGORY_ERROR'));
			return false;
		}
	}
```

**Comment by richard67** [comment]
@HLeithner Is not my PR, is @infograf768

**Comment by infograf768** [comment]
will test. No idea what happens btw in general when that plugin is disabled.

**Comment by infograf768** [comment]
@richard67 @HLeithner 
used plugin instead. Please test. 
(The lang string has to remain in com_content.ini as that plugin is not loading its lang files in the page concerned.)

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on b04c35298dfa2c87ca24a88a342426880a3ece8b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25798">issues.joomla.org/tracker/joomla-cms/25798</a>.</sub>

**Comment by richard67** [comment]
@franz-wohlkoenig Could you test again?

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on b04c35298dfa2c87ca24a88a342426880a3ece8b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25798">issues.joomla.org/tracker/joomla-cms/25798</a>.</sub>

**Comment by HLeithner** [comment]
Thanks for catching this error.

**Pull request merged**